spring-integration-core/src/main/java/org/springframework/integration/gateway/GatewayMethodInboundMessageMapper.java

```diff
@@ -298,6 +297,10 @@ private static Expression parsePayloadExpression(Method method) {
 					headersToMap != null
 							? new HashMap<>(headersToMap)
 							: new HashMap<>();
+
+			headersToPopulate.put("gatewayMethod", holder.getMethod());
+			headersToPopulate.put("gatewayArgs", holder.getArgs());
```

Comment:
We should probably add a mechanism to suppress these so the previous behavior can be retained if needed, especially for "classic" (non-DSL) gateways.

e.g. integration property `spring.integration.gateway.suppress.default.headers.interfaces=`

Reply:
??? This is just about two extra header in the message. There is no any harm in logic. More over you are going to simplify many places in your config with those extra configurations for `method` header and so on.
It is already there, so config is simpler ever if that is non-DSL one.
This is what we agreed in private discussion. 
In other places we populate source-based headers unconditionally, so what is the problem here?
If you don't need then downsrteam, you can just `filter` them!

Reply:
My real concern is adding overhead to serialization with large arguments where the serialization may be so far downstream that these arguments are no longer relevant. Of course `Method` is not `Serializable` so that won't be an issue.

Another option would be to set a boolean on the GPFB in `IntegrationFlows.from(Class<?> serviceInterface, String beanName)` so we don't get these for "legacy" gateways.

Reply:
OK. How about a `boolean mapInternalHeaders` option on a `MessagingGateway` and `<gateway>` with `false` by default?
It is indeed going to be `true` for `IntegrationFlows.from(Class<?> serviceInterface, String beanName)`, since this is only the place where we may not have access to annotation.

Reply:
Sounds good.